### PR TITLE
Support for BigQuery parameterization

### PR DESCRIPTION
**Pull request by yebrahim**
This PR adds support for query parameterization by passing a dictionary of parameters to BigQuery. parameters are inserted into execute calls, and look like:
```
%%bq execute --query myquery
parameters:
- name: param1
  type: STRING
  value: myvalue
- name: param2
...
```
Note the format of the schema is simpler than the one used by BigQuery, and is parsed into that format before sending to the service. The PR adds a jsonschema schema to (superficially) validate the input format before sending it to BigQuery.

Note this only adds support for named parameters. Positional parameters aren't supported yet.

### CODE REVIEW

File: google/datalab/bigquery/commands/_bigquery.py [hunk 1]
```diff
@@ -155,6 +155,7 @@ def _create_query_subparser(parser):
   query_parser.add_argument('--datasources', help='List of external datasources to reference in the query body',
                             nargs='+')
   query_parser.add_argument('--subqueries', help='List of subqueries to reference in the query body', nargs='+')
+  query_parser.add_argument('--parameters', help='List of subqueries to reference in the query body', nargs='+')
```
Comment (parthea) [inline google/datalab/bigquery/commands/_bigquery.py]: s/List of subqueries/List of parameters/

Comment (nikhilk) [inline google/datalab/bigquery/commands/_bigquery.py]: I think parameters realistically need to go into the cell body.

1. They won't necessarily fit on the one line.
2. You need to support both positional as well as named parameters. Commandline won't help with latter.

Comment (yebrahim) [inline google/datalab/bigquery/commands/_bigquery.py]: @parthea thanks for catching that

@nikhilk We also need the query body to go in the cell. One solution I see is using yaml syntax like we do in other magics, something like:
```
%%bq query --name myquery
- query: SELECT * FROM `mydataset.mytable WHERE field1 = @param1 AND field2 >= @param2
- parameters:
  - name: param1
    type: STRING
    value: :value
  - name: param2
    type: INT64
    value: 50
```
However, I feel like putting more things in the cell body is making writing magics more difficult, it's not really much easier than using the module API anymore. It also takes away from the spirit of writing SQL freely without worrying about delimiters and line breaks, but using the above syntax will push you to think about that. Do you have a better idea?

I think using a dictionary (or list for positional?) to contain parameters is a better experience than putting everything in the cell body.

**Comment by nikhilk** [comment]
I would expect parameters to be specified in bq execute. Not when defining
a query in bq query (same holds true in API).

Execute is where the binding of parameters to values happens. So in that
sense bq query is just SQL, and bq execute is just yaml syntax for
parameters.

**Comment by yebrahim** [comment]
@nikhilk sounds good, I'll switch to passing parameters to execute.

**Comment by yebrahim** [comment]
This is now ready for review.

**Comment by nikhilk** [comment]
Suggest:

```
%bq execute  ....
parameters
- name: p1
  type: STRING
  value: some_value
```

**Pull request merged**